src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py

```diff
@@ -504,8 +502,22 @@ def _validate_vmss_create_subnet(namespace):
             namespace.subnet_address_prefix = '{}/{}'.format(cidr, i)
 
         if namespace.app_gateway_type and namespace.app_gateway_subnet_address_prefix is None:
-            raise CLIError('Must specify --gateway-subnet-address-prefix to create an '
-                           'application gateway.')
+            namespace.app_gateway_subnet_address_prefix = _get_next_subnet_addr_suffix(
+                namespace.subnet_address_prefix, 24)
+
+
+def _get_next_subnet_addr_suffix(subnet_cidr, new_mask):
+    def _convert_to_int(address, bit_mask_len):
+        a, b, c, d = [int(x) for x in address.split('.')]
+        result = '{0:08b}{1:08b}{2:08b}{3:08b}'.format(a, b, c, d)
+        return int(result[:-bit_mask_len], 2)
+    ip_address, mask = subnet_cidr.split('/')
+    bit_mask_len = 32 - int(mask)
+    candidate_int = _convert_to_int(ip_address, bit_mask_len) + 1
```

Comment:
Since we are doing this automagically for the user, we should verify that this subnet will still fall within the VNET range, and if would fall outside, then we should retry this by getting a subnet by decrementing rather than incrementing. If both of those fall outside, then the VNET it too small for both subnets and we should error.

Otherwise it looks like someone could specify the scaleset subnet at the very end of the VNET range, and then it tries to create the AppGateway subnet by incrementing and this will cause an error. Okay if a user does that manually, but if we are doing it for them it is bad on us. :grin: